README.md
# himawari-urls

Get URLs for Himawari 8 image tiles based on a given date.

[![npm][npm-image]][npm-url]
[![travis][travis-image]][travis-url]
[![standard][standard-image]][standard-url]

[npm-image]: https://img.shields.io/npm/v/himawari-urls.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/himawari-urls
[travis-image]: https://img.shields.io/travis/ungoldman/himawari-urls.svg?style=flat-square
[travis-url]: https://travis-ci.org/ungoldman/himawari-urls
[standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square
[standard-url]: http://npm.im/standard

## Install

```
npm install himawari-urls
```

## Usage

```js
var himawariURLs = require('himawari-urls')

himawariURLs({
  date: 'latest',   // default: 'latest' (must be Date object, timestamp, or 'latest')
  infrared: true,   // default: false
  zoom: 1,          // default: 2
  timeout: 20000    // default: 10000 (milliseconds)
}, function (err, tiles) {
  if (err) console.error(err)
  tiles.map(function (t) { console.log(t) })
})

// { name: '0_0.png',
//  url: 'http://himawari8-dl.nict.go.jp/himawari8/img/INFRARED_FULL/1d/550/2016/02/10/192000_0_0.png',
//  x: 0,
//  y: 0 }
```

## Contributing

Contributions welcome! Please read the [contributing guidelines](CONTRIBUTING.md) first.

## Acknowledgment

All source code in `index.js` has been adapted from [jakiestfu](https://github.com/jakiestfu)'s work in [himawari.js](https://github.com/jakiestfu/himawari.js).

## See Also

- [hi8](https://github.com/ungoldman/hi8): See Earth from Himawari-8 on your desktop every 10 minutes.
- [himawari-bg](https://github.com/ungoldman/himawari-bg): Set the latest image from Himawari 8 as your desktop background.
- [himawari-history](https://github.com/ungoldman/himawari-history): Download all Himawari 8 images in a date range.

## License

Original and modified work is released under the [MIT](LICENSE.md) license.


CHANGELOG.md
# himawari-urls change log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.2
* Update HTTP request base url to HTTPS (#2)

## 1.0.1
* bump outdated for security warning

## 1.0.0
* better internal defaults management
* allow function to be called without options
* update repo URL
* update dependencies

## 1.0.0-alpha.2
* add timeout option

## 1.0.0-alpha.1
* update license to MIT and give jakiestfu more credit
* readme updates

## 1.0.0-alpha
* engage


test/index.js
var test = require('tape')
var moment = require('moment')
var lastWeek = new Date(moment().subtract(7, 'days').format())
var himawariURLs = require('../')

test('Fetch latest URLs', function (t) {
  himawariURLs({
    date: 'latest'
  }, function (err, tiles) {
    t.error(err, 'no error')
    t.ok(tiles, 'tiles object exists')
    t.equals(tiles.length, 1, 'contains 1 object')
    t.ok(tiles[0].name, 'tile 0 contains name property: ' + tiles[0].name)
    t.ok(tiles[0].url, 'tile 0 contains url property: ' + tiles[0].url)
    t.ok(!isNaN(tiles[0].x), 'tile 0 contains x property: ' + tiles[0].x)
    t.ok(!isNaN(tiles[0].y), 'tile 0 contains y property: ' + tiles[0].y)
    t.end()
  })
})

test('Fetch URLs from last week at zoom 2', function (t) {
  himawariURLs({
    date: lastWeek,
    zoom: 2
  }, function (err, tiles) {
    t.error(err, 'no error')
    t.ok(tiles, 'tiles object exists')
    t.equals(tiles.length, 16, 'contains 16 objects')
    t.ok(tiles[15].name, 'tile 16 contains name property: ' + tiles[15].name)
    t.ok(tiles[15].url, 'tile 16 contains url property: ' + tiles[15].url)
    t.ok(!isNaN(tiles[15].x), 'tile 16 contains x property: ' + tiles[15].x)
    t.ok(!isNaN(tiles[15].x), 'tile 16 contains y property: ' + tiles[15].y)
    t.end()
  })
})

test('Fetch infared URLs', function (t) {
  himawariURLs({
    date: 'latest',
    infrared: true
  }, function (err, tiles) {
    t.error(err, 'no error')
    t.ok(tiles, 'tiles object exists')
    t.equals(tiles.length, 1, 'contains 1 object')
    t.ok(tiles[0].name, 'tile 0 contains name property: ' + tiles[0].name)
    t.ok(tiles[0].url, 'tile 0 contains url property: ' + tiles[0].url)
    t.ok(!isNaN(tiles[0].x), 'tile 0 contains x property: ' + tiles[0].x)
    t.ok(!isNaN(tiles[0].y), 'tile 0 contains y property: ' + tiles[0].y)
    t.end()
  })
})

test('Fetch URLs from last week at zoom 2', function (t) {
  himawariURLs({
    date: lastWeek,
    infrared: true,
    zoom: 2
  }, function (err, tiles) {
    t.error(err, 'no error')
    t.ok(tiles, 'tiles object exists')
    t.equals(tiles.length, 16, 'contains 16 objects')
    t.ok(tiles[15].name, 'tile 16 contains name property: ' + tiles[15].name)
    t.ok(tiles[15].url, 'tile 16 contains url property: ' + tiles[15].url)
    t.ok(!isNaN(tiles[15].x), 'tile 16 contains x property: ' + tiles[15].x)
    t.ok(!isNaN(tiles[15].y), 'tile 16 contains y property: ' + tiles[15].y)
    t.end()
  })
})


index.js
var moment = require('moment')
var request = require('request')
var extend = require('deep-extend')

var BASE_URL = 'https://himawari8-dl.nict.go.jp/himawari8/img/'
var INFRARED = 'INFRARED_FULL'
var VISIBLE_LIGHT = 'D531106'
var WIDTH = 550
var LEVEL = [
  '1d',
  '4d',
  '8d',
  '16d',
  '20d'
]

module.exports = himawariURLs

/**
 * Returns an array of objects containing URLs and metadata
 * for Himawari 8 image tiles based on a given date.
 * Options:
 * - date: Date object, Date string, or 'latest'
 * - infrared: boolean (optional)
 * - zoom: number (default: 1)
 *
 * @param  {Object}       opts      date: Date|String, infrared: Boolean
 * @param  {Function}     callback  The function to be called when URLs are ready
 */
function himawariURLs (opts, callback) {
  // allow for options to be omitted and callback to be first argument
  if (!callback && typeof opts === 'function') {
    callback = opts
    opts = {}
  } else {
    opts = opts || {}
  }

  if (!callback) callback = function () {}

  var options = extend({
    date: 'latest',
    infrared: false,
    timeout: 5000,
    zoom: 1
  }, opts)

  var baseURL = getBaseURL(options.infrared)

  resolveDate({
    base_url: baseURL,
    date: options.date,
    timeout: options.timeout
  }, function (err, now) {
    if (err) return callback(err)

    // Normalize our date
    normalizeDate(now)

    // Define some image parameters
    var level = options.zoom ? LEVEL[options.zoom - 1] : LEVEL[0]
    var blocks = parseInt(level.replace(/[a-zA-Z]/g, ''), 10)
    var time = moment(now).format('HHmmss')
    var year = moment(now).format('YYYY')
    var month = moment(now).format('MM')
    var day = moment(now).format('DD')

    // compose URL
    var tilesURL = [baseURL, level, WIDTH, year, month, day, time].join('/')
    var tiles = []

    for (var x = 0; x < blocks; x++) {
      for (var y = 0; y < blocks; y++) {
        var name = x + '_' + y + '.png'
        var url = tilesURL + '_' + name

        tiles.push({
          name: name,
          url: url,
          x: x,
          y: y
        })
      }
    }

    callback(null, tiles)
  })
}

/**
 * Returns base URL for Himawari-8 images
 * @param   {Boolean}   infrared  returns base URL for infrared images if true
 * @returns {String}              full base URL
 */
function getBaseURL (infrared) {
  var url = BASE_URL
  if (infrared) url += INFRARED
  else url += VISIBLE_LIGHT
  return url
}

/**
 * Takes a Date and normalizes it to ten minute intervals.
 * @param   {Date}  input   Date object
 * @returns {Date}          normalized Date object
 */
function normalizeDate (date) {
  date.setMinutes(date.getMinutes() - (date.getMinutes() % 10))
  date.setSeconds(0)
  return date
}

/**
 * Takes an input, either a date object, a date timestamp, or the string 'latest'
 * and resolves to a native Date object. Requires base URL to resolve date based
 * on latest date retrieved from Himawari 8 server. Optionally override default
 * 10 second timeout.
 *
 * @param  {Object}       options   date (Date|String), uri (String), timeout (Number)
 * @param  {Function}     callback  The function to be called when date is resolved
 */
function resolveDate (options, callback) {
  var input = options.date
  var date = input

  // If provided a date string
  if (typeof input === 'string' && input !== 'latest') {
    date = new Date(input)
  }

  // If provided a date object
  if (moment.isDate(date)) return callback(null, date)

  // If provided 'latest'
  if (input === 'latest') {
    return request({
      method: 'GET',
      uri: options.base_url + '/latest.json',
      timeout: options.timeout || 10000
    }, function (err, res) {
      if (err) return callback(err)

      try {
        date = new Date(JSON.parse(res.body).date)
      } catch (e) {
        date = new Date()
      }

      return callback(null, date)
    })
  }

  // Invalid string provided, return new Date
  return callback(new Date())
}
